src/schema/posts.ts

```diff
@@ -1353,20 +1361,42 @@ export const resolvers: IResolvers<any, Context> = {
     ): Promise<GQLEmptyResponse> => updatePromoteToPublicFlag(ctx, id, null),
     updatePinPost: async (
       _,
-      { id, pinned }: PinPostArgs,
+      { id, pinned, nextPostId, prevPostId }: PinPostArgs,
       ctx: Context,
     ): Promise<GQLEmptyResponse> => {
       await ctx.con.transaction(async (manager) => {
         const repo = manager.getRepository(Post);
         const post = await repo.findOneBy({ id });
+        const args = { pinnedAt: pinned ? new Date() : null };
 
         await ensureSourcePermissions(
           ctx,
           post.sourceId,
           SourcePermissions.PostPin,
         );
 
-        await repo.update({ id }, { pinnedAt: pinned ? new Date() : null });
+        if (post.pinnedAt && (nextPostId || prevPostId)) {
+          const swapWithPost = await repo.findOneBy({
+            id: nextPostId || prevPostId,
+          });
+          const swapPinnedTime = swapWithPost.pinnedAt.getTime();
+          const randomizedIncrement = Math.floor(Math.random() * 10) * 1000;
+
+          console.log(
+            { swapPinnedTime, post: post.pinnedAt.getTime() },
+            swapPinnedTime > post.pinnedAt.getTime(),
+          );
+
+          if (nextPostId) {
+            args.pinnedAt = new Date(swapPinnedTime - 10 * 1000);
+          }
+
+          if (prevPostId) {
+            args.pinnedAt = new Date(swapPinnedTime + 10 * 1000);
+          }
+        }
```

Comment:
I will extract this into a separate mutation, but I'd like to get some feedback first